[jit] use save/load for emitFunctionHook

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22504 [WIP] classtypes hold weak_ptr
* #22221 [jit] _script_compile and _script_class_compile add to the python CU
* #22207 [jit] refactor self to be a class again
* #22206 [jit] Give functions qualified names
* **#22314 [jit] use save/load for emitFunctionHook**
* #22205 [jit] remove unused argument in import.cpp
* #22204 [jit] improvements to QualifiedName
* #22203 [jit] refactoring of module/object
* #22202 [jit] Make CompilationUnit own Functions
* #22502 [jit] make test context managers exception safe

1. Fix the `Function::save_to_buffer` so it actually works :O
2. Use the same save/load path for both emitFunctionHook and
emitModuleHook, eliminating the need to mess with CUs directly

Differential Revision: [D16121781](https://our.internmc.facebook.com/intern/diff/D16121781)

suo committed:
[jit] use save/load for emitFunctionHook

1. Fix the `Function::save_to_buffer` so it actually works :O
2. Use the same save/load path for both emitFunctionHook and
emitModuleHook, eliminating the need to mess with CUs directly

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -598,6 +598,7 @@ void initJitScriptBindings(PyObject* module) {
             std::ostringstream buf;
             Module module("__main__");
             addFunctionToModule(module, self);
+            module.save(buf, _extra_files);
```
lol

zdevito commented on test/jit_utils.py:
```diff
@@ -71,44 +71,17 @@ def _isHookExceptionOk(self, e):
-                    raise
-
-    def emitModuleHook(self, module):
+    def _compared_saved_loaded(self, m):
```
This will eventually have to be split apart again because saving a function will look different than saving a module with "new" serialization.

suo committed:
Update on "[jit] use save/load for emitFunctionHook"

[jit] use save/load for emitFunctionHook

1. Fix the `Function::save_to_buffer` so it actually works :O
2. Use the same save/load path for both emitFunctionHook and
emitModuleHook, eliminating the need to mess with CUs directly

gh-metadata: pytorch pytorch 22314 gh/suo/73/head

suo committed:
Update on "[jit] use save/load for emitFunctionHook"

[jit] use save/load for emitFunctionHook

1. Fix the `Function::save_to_buffer` so it actually works :O
2. Use the same save/load path for both emitFunctionHook and
emitModuleHook, eliminating the need to mess with CUs directly

gh-metadata: pytorch pytorch 22314 gh/suo/73/head
